app/Http/Controllers/ResultController.php

```diff
@@ -18,6 +20,80 @@ public function __construct()
         $this->middleware('auth:api');
     }
 
+    /**
+     * @param int $electionId
+     * @return Illuminate\Collection of all CandidateRanks associated with
+     * the election identified by the parameterized  id.
+     */
+    public function getCandidateRankCollection($electionId)
+    {
+        /*
+        SELECT 
+            candidate_ranks.elector_id,
+            candidate_ranks.rank,
+            candidate_ranks.candidate_id
+        FROM
+            candidate_ranks
+                JOIN
+            candidates ON candidates.id = candidate_ranks.candidate_id
+        WHERE
+            candidates.election_id = '4'
+        ORDER BY elector_id , rank;
+        */
+        $candidateRanks = CandidateRank::
+            join('candidates', 'candidate_ranks.candidate_id', '=', 'candidates.id')
+            ->where('candidates.election_id', '=', $electionId)
+            ->orderBy('elector_id', 'asc')
+            ->orderBy('rank', 'asc')
+            ->get();
+        return $candidateRanks;
+    }
+
+    /**
+     * @param Illuminate\Database\Eloquent\Relations\Collection of App\CandidateRank
+     * @return array of arrays of arrays. The arrays are grouped at the outermost level
+     * on elector id. The arrays are grouped at the next level by rank. Array entries at this level are ordered in
+     * ascending rank. The innermost arrays are associative arrays that contain CandidateRank attributes.
+     */
+    public function groupRankingsByElectorAndRank($candidateRanks)
+    {
+        $candidateRanksGroupedByElector = $candidateRanks->mapToGroups(function($candidateRank){
+            $value = $candidateRank;
+            $key = $candidateRank->getAttributeValue('elector_id');
+            return [ $key => $value ];
+        });
+
+        $candidateRanksGroupedByElectorAndRank = $candidateRanksGroupedByElector->map(function($candidateRanksFromOneElector){
+            return $candidateRanksFromOneElector->mapToGroups(function($candidateRank){
+                $value = $candidateRank;
+                $key = $candidateRank->getAttributeValue('rank');
+                return [ $key => $value ];
+            });
+        })->toArray();
+        return $candidateRanksGroupedByElectorAndRank;
+    }
+
+    /**
+     * @param array of arrays of arrays as output by $this->groupRankingsByElectorAndRank
+     * @return array of NBallots
+     */
+    public function buildNBallots($candidateRanksGroupedByElectorAndRank)
```

Comment:
Do we need the NBallot concept?  There are N factorial different combos, so dedup at that level seems of limited value...

Reply:
Good point. I will add a task for simplifying.

Reply:
https://github.com/pivot-libre/tideman/issues/56